src/Domain/Structure/OrientationMap.cpp

```diff
@@ -126,6 +129,43 @@ std::ostream& operator<<(std::ostream& os,
   return os;
 }
 
-template class OrientationMap<1>;
-template class OrientationMap<2>;
-template class OrientationMap<3>;
+template <size_t VolumeDim, typename T>
+std::array<tt::remove_cvref_wrap_t<T>, VolumeDim> discrete_rotation(
+    const OrientationMap<VolumeDim>& rotation,
+    std::array<T, VolumeDim> source_coords) noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  std::array<ReturnType, VolumeDim> new_coords{};
+  for (size_t i = 0; i < VolumeDim; i++) {
+    const auto new_direction = rotation(Direction<VolumeDim>(i, Side::Upper));
+    gsl::at(new_coords, i) =
+        std::move(gsl::at(source_coords, new_direction.dimension()));
+    if (new_direction.side() != Side::Upper) {
+      gsl::at(new_coords, i) *= -1.0;
+    }
+  }
+  return new_coords;
+}
+
+#define DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define DTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
+
+#define INSTANTIATION(r, data)                                         \
+  template std::array<tt::remove_cvref_wrap_t<DTYPE(data)>, DIM(data)> \
+  discrete_rotation<DIM(data), DTYPE(data)>(                           \
+      const OrientationMap<DIM(data)>& rotation,                       \
+      std::array<DTYPE(data), DIM(data)> source_coords) noexcept;
+
+GENERATE_INSTANTIATIONS(INSTANTIATION, (1ul, 2ul, 3),
```

Comment:
For my understanding: why are some of these marked `ul`, we don't seem to need this in most places in the repo.

Also, why the inconsistent marking?

Reply:
legacy, I'll remove.

What do you mean by `marking`, sorry?

Reply:
I mean suffixing (why are some `Xul`, and another just `X`)